nomad/plan_apply_node_tracker.go

```diff
@@ -0,0 +1,203 @@
+package nomad
+
+import (
+	"fmt"
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/hashicorp/nomad/helper"
+	"golang.org/x/time/rate"
+)
+
+type BadNodeTracker interface {
+	IsBad(string) bool
+	Add(string)
+	EmitStats(time.Duration, <-chan struct{})
+}
+
+// NoopBadNodeTracker is a no-op implementation of bad node tracker that is
+// used when tracking is disabled.
+type NoopBadNodeTracker struct{}
+
+func (n *NoopBadNodeTracker) Add(string)                               {}
+func (n *NoopBadNodeTracker) EmitStats(time.Duration, <-chan struct{}) {}
+func (n *NoopBadNodeTracker) IsBad(string) bool {
+	return false
+}
+
+// CachedBadNodeTracker keeps a record of nodes marked as bad by the plan
+// applier in a LRU cache.
+//
+// It takes a time window and a threshold value. Plan rejections for a node
+// will be registered with its timestamp. If the number of rejections within
+// the time window is greater than the threshold the node is reported as bad.
+//
+// The tracker uses a fixed size cache that evicts old entries based on access
+// frequency and recency.
+type CachedBadNodeTracker struct {
+	logger    hclog.Logger
+	cache     *lru.TwoQueueCache
+	limiter   *rate.Limiter
+	window    time.Duration
+	threshold int
+}
+
+type CachedBadNodeTrackerConfig struct {
+	CacheSize int
+	RateLimit float64
+	BurstSize int
+	Window    time.Duration
+	Threshold int
+}
+
+func DefaultCachedBadNodeTrackerConfig() CachedBadNodeTrackerConfig {
+	return CachedBadNodeTrackerConfig{
+		CacheSize: 50,
+
+		// Limit marking 5 nodes per 30min as ineligible with an initial
+		// burst of 10 nodes.
+		RateLimit: 5.0 / (30 * 60),
+		BurstSize: 10,
+
+		// Consider a node as bad if it is added more than 100 times in a 5min
+		// window period.
+		Window:    5 * time.Minute,
+		Threshold: 100,
+	}
+}
+
+// NewCachedBadNodeTracker returns a new CachedBadNodeTracker.
+func NewCachedBadNodeTracker(logger hclog.Logger, config CachedBadNodeTrackerConfig) (*CachedBadNodeTracker, error) {
+	log := logger.Named("bad_node_tracker").
+		With("window", config.Window).
+		With("threshold", config.Threshold)
+
+	cache, err := lru.New2Q(config.CacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new bad node tracker: %v", err)
+	}
+
+	return &CachedBadNodeTracker{
+		logger:    log,
+		cache:     cache,
+		limiter:   rate.NewLimiter(rate.Limit(config.RateLimit), config.BurstSize),
+		window:    config.Window,
+		threshold: config.Threshold,
+	}, nil
+}
+
+// IsBad returns true if the node has more rejections than the threshold within
+// the time window.
+func (c *CachedBadNodeTracker) IsBad(nodeID string) bool {
+	logger := c.logger.With("node_id", nodeID)
+	logger.Debug("checking if node is bad")
+
+	value, ok := c.cache.Get(nodeID)
+	if !ok {
+		logger.Debug("node not in cache")
+		return false
```

Comment:
It looks like we should never hit this because calling `Add` first is an invariant of the design. We could return an error here instead, or even better change the construction of `Add` and `IsBad` so that we both add and check in the same call in `plan_apply.go`

Reply:
Yeah that makes sense. I modified the interface so it only exposes `Add` and `EmitMetrics`. `isBad` now takes the object in the cache that `Add` created or retrieved, so there's no need for this check anymore 👍